content/2023-06-02-shrinking-binaries.rst

```diff
@@ -0,0 +1,358 @@
+Shrinking a Shared Library
+##########################
+
+:date: 2023-06-02
+:category: mozilla
+:lang: en
+:authors: Serge Guelton
+:summary: What is the effect of different compiler flags on the size of a shared
+          libary? Let's explore!
+
+Recently, I've submitted a `patch
+<https://phabricator.services.mozilla.com/D179806>`_ that shaves ~2.5MB on one
+of the core Firefox library, ``libxul.so``. The patch is stupidly simple, but
+it's a good opportunity to dive into some techniques to shrink binary size.
+
+Generic Approach
+================
+
+We'll use the ` ``libz`` <https://github.com/madler/zlib>`_ library
+from the official `zlib <https://zlib.net/>`_ as an example program to build. In
+this series, we use revision **1.2.13**, sha1 **04f42ceca40f73e2978b50e93806c2a18c1281fc**.
+
+The compiler used is clang **14.0.5**, the linker used is lld.
+
+Compilation Levels
+------------------
+
+Let's start with the obvious approach and explore the effect of the various
+optimization levels ``-O0``, ``-O1``, ``-Oz``, ``-O2`` and ``-O3``.
+
+The ``size`` program *list section sizes and total size of binary files*, that's
+perfectly suited to our needs. We will dive into its detailed output later.
+
+.. code-block:: sh
+
+    $ LDFLAGS=-fuse-ld=lld CFLAGS=-O0 CC=clang ../configure && make && size -A libz.so
+    [...]
+    Total                   141907
+
+.. code-block:: sh
+
+    $ LDFLAGS=-fuse-ld=lld CFLAGS=-O1 CC=clang ../configure && make && size -A libz.so
+    [...]
+    Total                    88463
+
+.. code-block:: sh
+
+    $ LDFLAGS=-fuse-ld=lld CFLAGS=-O2 CC=clang ../configure && make && size -A libz.so
+    [...]
+    Total                    99831
+
+.. code-block:: sh
+
+    $ LDFLAGS=-fuse-ld=lld CFLAGS=-Oz CC=clang ../configure && make && size -A libz.so
+    [...]
+    Total                    81305
+
+.. code-block:: sh
+
+    $ LDFLAGS=-fuse-ld=lld CFLAGS=-O3 CC=clang ../configure && make && size -A libz.so
+    [...]
+    Total                   101943
+
+Unsurprisingly ``-Oz`` yields the smaller binaries. ``-O3`` trades size for
+performance, and ``-O0`` performs a direct translation which uses a lot of
+space.
```

Comment:
What about `-02` v.s. `-01`?

Also, I [believe ](https://clang.llvm.org/docs/CommandGuide/clang.html#cmdoption-o0) `-Oz` is `-O2` + extra options. Any chance that `-01` plus the same options can give smaller binaries (with some perf tradeoff)?